### PR TITLE
Implement Phase 3

### PR DESCRIPTION
**Pull request by mellowcroc**
Serialize proof for Starknet (L2)

### CODE REVIEW

File: src/serialize.rs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    let (config, public_input, unsent_commitment, witness) = parse_proof_file(&proof_file)?;
+
+    let proof = chain!(
+        config.into_iter(),
+        public_input.into_iter(),
+        unsent_commitment.into_iter(),
+        witness.into_iter()
+    );
+
+    let calldata = chain!(proof, std::iter::once(Felt252::from(1)));
```
Comment (rot256) [inline src/serialize.rs:23]: What's the magic "1" constant?

File: src/serialize.rs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    fs::write(args.output.clone(), calldata_string)?;
+    Ok(())
+}
+
+fn parse_proof_file(proof_file: &Path) -> Result<(VecFelt252, VecFelt252, VecFelt252, VecFelt252)> {
+    let proof_file_content = std::fs::read_to_string(proof_file)?;
+    let parsed = parse(proof_file_content)?;
+    Ok((
+        serde_json::from_str(&parsed.config.to_string())?,
```
Comment (rot256) [inline src/serialize.rs:38]: Looks fine.

If you don't want to do: serialize -> deserialize you could also parse the Felt252 from each Expr::Value in the Exprs.

**Review by rot256 — APPROVED** [review]
Looks great!

I only had a couple, very minor, comments.
Sorry it took so long...

File: src/serialize.rs (hunk 2 above)
Comment (mellowcroc) [reply src/serialize.rs:54]: Good point--will take advantage of the existing implementations now though :)

File: src/serialize.rs (hunk 1 above)
Comment (mellowcroc) [reply src/serialize.rs]: Nice catch, added the code for specifying this. (Cairo 1)

**Pull request merged**